cpp/cycles_module/algorithm/cycles.hpp

```diff
@@ -0,0 +1,61 @@
+#include <set>
+
+#include <mg_graph.hpp>
+
+namespace cycles_util {
+
+/// Simple struct that keeps the state of nodes in algorithms that rely on
+/// DFS traversal.
+struct NodeState {
+  std::vector<bool> visited;
+  std::vector<std::uint64_t> discovery, depth;
+  std::vector<std::uint64_t> parent;
+
+  explicit NodeState(size_t number_of_nodes);
+
+  void SetVisited(std::uint64_t node_id);
+
+  bool IsVisited(std::uint64_t node_id);
+
+  void SetParent(std::uint64_t parent_id, std::uint64_t node_id);
+
+  std::uint64_t GetParent(std::uint64_t node_id);
+
+  void SetDepth(std::uint64_t node_id, std::uint64_t node_depth);
+
+  std::uint64_t GetDepth(std::uint64_t node_id);
```

Comment:
In addition to this. Is it necessary to have an implementation in another file, once when implementation is trivial as is in this example?

Reply:
Yeah, sometimes it's annoying to put a definition into a cpp but I would still do that with the exception of the methods whose implementation are really obvious in general. Example of this would be getter functions 
```
struct Foo {
 const auto &field() const { return field_; }
 private:
  int field_;
}; 
```
It's obvious that it's a getter function, they will always return a direct reference to the field and you can use `const auto &` for the return type.

For all of this, it's not that obvious what the implementation is. What if you want to add checks later on for the ids? Every time you add something new to the header forces recompilation of every file that includes that header. If you kept it into cpp you would only have to recompile that one file.